src/components/shared/MarkerTooltipContents.js

```diff
@@ -37,19 +39,80 @@ function _markerDetail<T: NotVoidOrNull>(
   fn: T => string = String
 ): React.Node {
   return [
-    <div className="tooltipLabel" key="{key}">
+    <div className="tooltipLabel" key={key}>
       {label}:
     </div>,
     fn(value),
   ];
 }
 
+type PhaseTimeTuple = { name: string, time: Microseconds };
+
+function _markerDetailPhase(p: PhaseTimeTuple): React.Node {
+  return _markerDetail(
+    'gcphase' + p.name,
+    'Phase ' + p.name,
+    p.time,
+    t => formatNumber(t / 1000) + 'ms'
+  );
+}
+
+function _makePhaseTimesArray(
+  phases: PhaseTimes<Microseconds>
+): Array<PhaseTimeTuple> {
+  const array = [];
+  for (const phase in phases) {
+    if (phase !== 'Total') {
+      array.push({ name: phase, time: phases[phase] });
+    }
+  }
+  return array;
+}
+
+function _leafPhase(phases: Array<PhaseTimeTuple>, name: string): boolean {
+  for (let i = 0; i < phases.length; i++) {
+    const pt = phases[i];
+    if (pt.name === name) {
+      // Don't consider itself.
+      continue;
+    }
+    if (pt.name.startsWith(name)) {
+      // Name is not a leaf
+      return false;
+    }
+  }
+  // name is a leaf.
+  return true;
+}
+
+function _filterInterestingPhaseTimes(
+  raw_phases: PhaseTimes<Microseconds>,
+  num_select: number
+): Array<PhaseTimeTuple> {
+  let phase_times = _makePhaseTimesArray(raw_phases);
+
+  // Select only the leaf phases.
+  phase_times = phase_times.filter(pt => _leafPhase(phase_times, pt.name));
+
+  // Find the top N phases.
+  const threshold = Math.min(
+    ...phase_times
+      .map(pt => pt.time)
+      .sort((a, b) => Number(b) - Number(a))
+      .slice(0, num_select + 1)
+  );
+  phase_times = phase_times.filter(pt => pt.time > threshold);
+
```

Comment:
why using `Math.min` if the array is already sorted? Isn't the min the last element?
In JS it's painful to get the last element, so it's also possible to sort in ascending direction, slice the end instead of the start, and get the first element. Also because we can avoid `Math.min` we can also skip the `map` step.

But, actually, because we sort the array, and slice the n best, why filtering again at the end ? Don't we have the elements we want already?

Reply:
A proposal:
```javascript
// return the top N phases in descending order.
return phaseTimes
  .sort((a, b) => b.time - a.time)
  .slice(0, numSelect);
```

Reply:
I like your code, it's much simplier, However my code presents the phases in the order they're executed by the GC.  Therefore finding the threshold and filtering using it makes sense.  I'll add a comment.